src/trusted/service_runtime/nacl_syscall_common.c

```diff
@@ -1238,70 +1238,38 @@ int32_t NaClSysLseek(struct NaClAppThread *natp,
 
 int32_t NaClSysIoctl(struct NaClAppThread *natp,
                      int                  d,
-                     int                  request,
-                     void                 *arg) {
+                     unsigned long        request,
+                     void                 *arg_ptr) {
   struct NaClApp  *nap = natp->nap;
   int             retval = -NACL_ABI_EINVAL;
   uintptr_t       sysaddr;
-  struct NaClDesc *ndp;
   int             fd;
 
-  NaClLog(2, "Entered NaClSysIoctl(0x%08"NACL_PRIxPTR
+  NaClLog(2, "Cage %d Entered NaClSysIoctl(0x%08"NACL_PRIxPTR
            ", %d, %d, 0x%08"NACL_PRIxPTR")\n",
-           (uintptr_t)natp, d, request,
-           (uintptr_t)arg);
-  /*
-   * Note that NaClUserToSysAddrRange is not feasible right now, since
-   * the size of the arg argument depends on the request.  We do not
-   * have an enumeration of allowed ioctl requests yet.
-   *
-   * Furthermore, some requests take no arguments, so sysaddr might
-   * end up being kNaClBadAddress and that is perfectly okay.
-   */
-  sysaddr = NaClUserToSysAddr(nap, (uintptr_t) arg);
-  /*
-   ****************************************
-   * NOTE: sysaddr may be kNaClBadAddress *
-   ****************************************
-   */
-
-  fd = fd_cage_table[nap->cage_id][d];
+           nap->cage_id, (uintptr_t)natp, d, request,
+           (uintptr_t)arg_ptr);
 
-  /* check for closed fds */
-  if (fd < 0) {
-    retval = -NACL_ABI_EBADF;
-    goto cleanup;
+  sysaddr = NaClUserToSysAddr(nap, (uintptr_t) arg_ptr);
```

Comment:
I think you need to check whether it's actually supposed to be a pointer type here, i.e. for I_POP, I_FLUSH, etc. we don't want to convert the pointer.

Reply:
The I_POP seems to be an anomaly in requiring a 0, but I think it can be dealt with by simply passing a pointer to an integer 0 rather than restructuring the whole thing. I want to be agnostic about the input type on the GLIBC/IOCTL side as much as possible so that minimal change is ever necessary outside SafePosix.

Reply:
I_FLUSH however will require an integer, which will break on NaClUserToSysAddr

Reply:
Hm, I understand that. But again, we can solve it in glibc the same way (by creating pointers to those values) as we will create checks for the ones we support anyways (these calls are told to be unimplemented directly in GlibC right now so they do not reach this area.). 

Otherwise, if we truly want to pass values for these functions, we will have to explicitly separate different syscall types a lot in all 3 repos and we will have to deal with explicit casts to pointer types or int types.

Either way we go, such changes can be implemented if we add those syscalls as adding the checks won't do anything for now.

Reply:
Ah ok. I think as much as possible we should try to avoid changing the way arguments are handled in syscalls. I think a comment that this as is will not work for ioctls which take integers would be warranted though.